packages/doenetml-worker-rust/lib-rust-macros/src/test_data_query_types/test_data_query_types.rs

```diff
@@ -0,0 +1,106 @@
+use proc_macro2::TokenStream;
+
+use darling::{FromDeriveInput, FromField};
+use quote::{format_ident, quote};
+use syn::{spanned::Spanned, DeriveInput, Meta, Result};
+
+#[derive(Debug, FromDeriveInput)]
+#[darling(forward_attrs(owning_component))]
+struct FromDeriveInputStruct {
+    ident: syn::Ident,
+    data: darling::ast::Data<darling::util::Ignored, Field>,
+    // Because we want to allow `#[owning_component(CompName)]` instead of forcing `#[owning_component(name = CompName)]`,
+    // we have to grab the raw attrs directly.
+    attrs: Vec<syn::Attribute>,
+}
+
+#[derive(Debug, FromField)]
+#[darling(attributes(from_field))]
+struct Field {
+    // We don't care about field idents for now (only variant idents)
+    ident: Option<syn::Ident>,
+    ty: syn::Type,
+}
+
+pub fn test_data_query_types_derive(ts: TokenStream) -> Result<TokenStream> {
+    let ast: DeriveInput = syn::parse2(ts)?;
+    let parsed: FromDeriveInputStruct = FromDeriveInputStruct::from_derive_input(&ast)?;
+
+    let struct_name = parsed.ident.clone();
+    let component_name = component_name_from_attrs(&parsed.attrs).ok_or(syn::Error::new(
+        parsed.ident.span(),
+        "No `owning_component` attribute found. Please add #[owning_component(CompName)]",
+    ))?;
+
+    // We build up the vec in a loop so we can propagate errors
+    let mut variants = Vec::new();
+    for variant in parsed
+        .data
+        .take_struct()
+        .expect("Struct not found. Can only implement on structs")
+        .iter()
+    {
+        let variant_ident = variant.ident.clone().ok_or(syn::Error::new(
+            variant.ident.span(),
+            "Anonymous struct variants are not supported",
+        ))?;
+        let ty = variant.ty.clone();
+        variants.push((variant_ident, ty));
+    }
+
+    let type_entry = variants
+        .iter()
+        .map(|(variant_ident, ty)| {
+            let ident_str = variant_ident.to_string();
+            quote! {
+                (#ident_str, <#ty>::PROP_VALUE_TYPE)
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let struct_name_str = struct_name.to_string();
+    let impls = quote! {
+        #[cfg(test)]
+        impl TestDataQueryTypes for #struct_name {
+            const _DECLARED_DATA_QUERY_TYPES: &'static [(&'static str, Option<PropValueType>)] = &[
+                #(#type_entry),*
+            ];
+
+            const _STRUCT_NAME: &'static str = #struct_name_str;
+        }
+    };
+
+    let test = quote! {
+        #[cfg(test)]
+        mod __types_test_auto_generated {
```

Comment:
If you didn't put the test in this separate module, what would be different?

Reply:
Shouldn't be a difference, but `#[cfg(test)]` will ensure the entire module is not compiled unless test=true. If we ever added another test, we can put it in the same module. Otherwise we would need to add multiple `#[cfg(test)]` lines.